Deserialize trait shouldn't be required for T in Subscriber<T>

`Deserialize` should only be required for types which are passed as params to rpc methods. A pubsub `subscribe` method accepts a `Subscriber<T>` and the visitor does not discriminate so assumes it requires the type bound.

Found when converting `substrate` to use `jsonrpc-derive`

```
Compiling substrate-rpc-servers v0.1.0 (/home/andrew/code/paritytech/substrate/core/rpc-servers)
error[E0277]: the trait bound `for<'de> <Block as sr_primitives::traits::Block>::Header: serde::Deserialize<'de>` is not satisfied
  --> core/rpc-servers/src/lib.rs:59:23
   |
59 |     io.extend_with(chain.to_delegate());
   |                          ^^^^^^^^^^^ the trait `for<'de> serde::Deserialize<'de>` is not implemented for `<Block as sr_primitives::traits::Block>::Header`
   |
   = help: consider adding a `where for<'de> <Block as sr_primitives::traits::Block>::Header: serde::Deserialize<'de>` bound
   = note: required because of the requirements on the impl of `sr_primitives::DeserializeOwned` for `<Block as sr_primitives::traits::Block>::Header`
```

Fix will be somewhere in the visitor here: https://github.com/paritytech/jsonrpc/blob/master/derive/src/to_delegate.rs#L360

--- a/derive/src/to_delegate.rs
+++ b/derive/src/to_delegate.rs
@@ -336,12 +336,14 @@ fn is_option_type(ty: &syn::Type) -> bool {
 }
 
 fn generate_where_clause_serialization_predicates(item_trait: &syn::ItemTrait) -> Vec<syn::WherePredicate> {
+	#[derive(Default)]
 	struct FindTyParams {
 		trait_generics: HashSet<syn::Ident>,
 		serialize_type_params: HashSet<syn::Ident>,
 		deserialize_type_params: HashSet<syn::Ident>,
 		visiting_return_type: bool,
 		visiting_fn_arg: bool,
+		visiting_subscriber_arg: bool,
 	}
 	impl<'ast> Visit<'ast> for FindTyParams {
 		fn visit_type_param(&mut self, ty_param: &'ast syn::TypeParam) {
@@ -356,10 +358,14 @@ fn generate_where_clause_serialization_predicates(item_trait: &syn::ItemTrait) -
 			if self.visiting_return_type && self.trait_generics.contains(&segment.ident) {
 				self.serialize_type_params.insert(segment.ident.clone());
 			}
-			if self.visiting_fn_arg && self.trait_generics.contains(&segment.ident) {
+			if self.visiting_fn_arg &&
+				self.trait_generics.contains(&segment.ident) &&
+				!self.visiting_subscriber_arg {
 				self.deserialize_type_params.insert(segment.ident.clone());
 			}
-			visit::visit_path_segment(self, segment)
+			self.visiting_subscriber_arg = self.visiting_fn_arg && segment.ident == SUBCRIBER_TYPE_IDENT;
+			visit::visit_path_segment(self, segment);
+			self.visiting_subscriber_arg = false;
 		}
 		fn visit_fn_arg(&mut self, arg: &'ast syn::FnArg) {
 			self.visiting_fn_arg = true;
@@ -367,13 +373,7 @@ fn generate_where_clause_serialization_predicates(item_trait: &syn::ItemTrait) -
 			self.visiting_fn_arg = false;
 		}
 	}
-	let mut visitor = FindTyParams {
-		visiting_return_type: false,
-		visiting_fn_arg: false,
-		trait_generics: HashSet::new(),
-		serialize_type_params: HashSet::new(),
-		deserialize_type_params: HashSet::new(),
-	};
+	let mut visitor = FindTyParams::default();
 	visitor.visit_item_trait(item_trait);
 
 	item_trait.generics

--- a/derive/tests/compiletests.rs
+++ b/derive/tests/compiletests.rs
@@ -9,7 +9,6 @@ fn run_mode(mode: &'static str) {
 	config.src_base = PathBuf::from(format!("tests/{}", mode));
 	config.target_rustcflags = Some(String::from(
 		"\
-			--edition=2018 \
 			-L ../target/debug/ \
 			-L ../target/debug/deps/ \
 	"));

--- a/derive/tests/compiletests.rs
+++ b/derive/tests/compiletests.rs
@@ -7,7 +7,12 @@ fn run_mode(mode: &'static str) {
 
 	config.mode = mode.parse().expect("Invalid mode");
 	config.src_base = PathBuf::from(format!("tests/{}", mode));
-	config.target_rustcflags = Some("-L ../target/debug/ -L ../target/debug/deps/".to_owned());
+	config.target_rustcflags = Some(String::from(
+		"\
+			--edition=2018 \
+			-L ../target/debug/ \
+			-L ../target/debug/deps/ \
+	"));
 	config.clean_rmeta(); // If your tests import the parent crate, this helps with E0464
 
 	compiletest::run_tests(&config);
@@ -16,4 +21,5 @@ fn run_mode(mode: &'static str) {
 #[test]
 fn compile_test() {
 	run_mode("ui");
+	run_mode("run-pass");
 }

--- a/derive/tests/run-pass/pubsub-subscription-type-without-deserialize.rs
+++ b/derive/tests/run-pass/pubsub-subscription-type-without-deserialize.rs
@@ -1,0 +1,29 @@
+#[macro_use]
+extern crate serde_derive;
+
+use jsonrpc_core::Result;
+use jsonrpc_derive::rpc;
+use jsonrpc_pubsub::{typed::Subscriber, SubscriptionId};
+
+// One way serialization
+#[derive(Serialize)]
+struct SerializeOnly {
+	foo: String,
+}
+
+#[rpc]
+pub trait Rpc {
+	type Metadata;
+
+	/// Hello subscription
+	#[pubsub(subscription = "hello", subscribe, name = "hello_subscribe", alias("hello_sub"))]
+	fn subscribe(&self, _: Self::Metadata, _: typed::Subscriber<SerializeOnly>, _: u64);
+
+	/// Unsubscribe from hello subscription.
+	#[pubsub(subscription = "hello", unsubscribe, name = "hello_unsubscribe")]
+	fn unsubscribe(&self, _: Option<Self::Metadata>, _: SubscriptionId) -> Result<bool>;
+}
+
+fn main() {
+	let _ = Rpc::to_delegate();
+}

--- a/derive/tests/run-pass/pubsub-subscription-type-without-deserialize.rs
+++ b/derive/tests/run-pass/pubsub-subscription-type-without-deserialize.rs
@@ -6,8 +6,22 @@ extern crate jsonrpc_pubsub;
 #[macro_use]
 extern crate jsonrpc_derive;
 
+use std::sync::Arc;
 use jsonrpc_core::Result;
-use jsonrpc_pubsub::{typed::Subscriber, SubscriptionId};
+use jsonrpc_pubsub::{typed::Subscriber, SubscriptionId, Session, PubSubHandler};
+
+#[rpc]
+pub trait Rpc<T> {
+	type Metadata;
+
+	/// Hello subscription
+	#[pubsub(subscription = "hello", subscribe, name = "hello_subscribe", alias("hello_sub"))]
+	fn subscribe(&self, _: Self::Metadata, _: Subscriber<T>);
+
+	/// Unsubscribe from hello subscription.
+	#[pubsub(subscription = "hello", unsubscribe, name = "hello_unsubscribe")]
+	fn unsubscribe(&self, _: Option<Self::Metadata>, _: SubscriptionId) -> Result<bool>;
+}
 
 // One way serialization
 #[derive(Serialize)]
@@ -15,19 +29,21 @@ struct SerializeOnly {
 	foo: String,
 }
 
-#[rpc]
-pub trait Rpc<SerializeOnly> {
-	type Metadata;
+struct RpcImpl;
+impl Rpc<SerializeOnly> for RpcImpl {
+	type Metadata = Arc<Session>;
 
-	/// Hello subscription
-	#[pubsub(subscription = "hello", subscribe, name = "hello_subscribe", alias("hello_sub"))]
-	fn subscribe(&self, _: Self::Metadata, _: Subscriber<SerializeOnly>);
+	fn subscribe(&self, _: Self::Metadata, _: Subscriber<SerializeOnly>) {
+		unimplemented!();
+	}
 
-	/// Unsubscribe from hello subscription.
-	#[pubsub(subscription = "hello", unsubscribe, name = "hello_unsubscribe")]
-	fn unsubscribe(&self, _: Option<Self::Metadata>, _: SubscriptionId) -> Result<bool>;
+	fn unsubscribe(&self, _: Option<Self::Metadata>, _: SubscriptionId) -> Result<bool> {
+		unimplemented!();
+	}
 }
 
 fn main() {
-	let _ = Rpc::to_delegate();
+	let mut io = PubSubHandler::default();
+	let rpc = RpcImpl;
+	io.extend_with(rpc.to_delegate());
 }

--- a/derive/tests/run-pass/pubsub-subscription-type-without-deserialize.rs
+++ b/derive/tests/run-pass/pubsub-subscription-type-without-deserialize.rs
@@ -1,8 +1,12 @@
+extern crate serde;
 #[macro_use]
 extern crate serde_derive;
+extern crate jsonrpc_core;
+extern crate jsonrpc_pubsub;
+#[macro_use]
+extern crate jsonrpc_derive;
 
 use jsonrpc_core::Result;
-use jsonrpc_derive::rpc;
 use jsonrpc_pubsub::{typed::Subscriber, SubscriptionId};
 
 // One way serialization
@@ -12,12 +16,12 @@ struct SerializeOnly {
 }
 
 #[rpc]
-pub trait Rpc {
+pub trait Rpc<SerializeOnly> {
 	type Metadata;
 
 	/// Hello subscription
 	#[pubsub(subscription = "hello", subscribe, name = "hello_subscribe", alias("hello_sub"))]
-	fn subscribe(&self, _: Self::Metadata, _: typed::Subscriber<SerializeOnly>, _: u64);
+	fn subscribe(&self, _: Self::Metadata, _: Subscriber<SerializeOnly>);
 
 	/// Unsubscribe from hello subscription.
 	#[pubsub(subscription = "hello", unsubscribe, name = "hello_unsubscribe")]

